Deactivate the tracking of gradients in sampling functions within OpInfos

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67833
* #68188
* #67996
* #68523
* __->__ #68522
* #69909
* #69908

Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

Differential Revision: [D33774420](https://our.internmc.facebook.com/intern/diff/D33774420)

lezcano committed:
Deactivate the tracking of gradines in  sampling functions within OpInfos

Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

Lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -655,7 +655,7 @@ def __init__(self,
         self.decorators = (*decorators, *skips)
 
-        self.sample_inputs_func = sample_inputs_func
+        self.sample_inputs_func = torch.no_grad()(sample_inputs_func)
```
nb. This is the important point of this PR. The rest of the PR just rewrites some tests for them to work as expected with this change and with the `non_contiguous` tests added in https://github.com/pytorch/pytorch/pull/67996

mruberry replied:
Comment here too

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

anjali411 reviewed (approved):
LGTM!

anjali411 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1946,8 +1941,7 @@ def sample_inputs_addmm(op_info, device, dtype, requires_grad, **kwargs):
                         args=(
-                            make_tensor(shape, device, dtype,
-                                        requires_grad=requires_grad).t().conj(),
+                            make_tensor(shape, device, dtype).mH.requires_grad_(requires_grad),
```
do you get an error if you don't set the requires grad? I know you left a note above to set requires grad after performing any transformations but I am worried because it's an easy thing to miss

Lezcano replied:
Yes, you do get an error. I needed to update all these functions for the tests to pass.

anjali411 replied:
maybe we should create helper functions for standard transformations that people can use. We can take care of retaining the `requires_grad` field in those helper functions then!

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -325,6 +325,10 @@ def _getattr_qual(obj, name, default=_NOTHING):
 #   Each SampleInput defines an "input", "args", "kwargs",
 #   an "output_process_fn_grad" function, the "broadcasts_input" bool and
 #   a "name".
+#   All the "sample_inputs" functions are invoked within a `torch.no_grad()`
```
Newline (new paragraph)

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1894,26 +1895,20 @@ def sample_inputs_t(op_info, device, dtype, requires_grad, **kwargs):
 
-    first_shape, second_shape = (S, M), (M, S)
-    sample_inputs = []
+    def make_arg_conj(size):
```
Wait what does this function do? Doesn't it need a conj()?

mruberry replied:
Follow-up question for @anjali411 -- should sample inputs be generating conjugated samples, or are those automatically generated now?

Lezcano replied:
Oh, sorry, I completely missed the `conj()`! Thanks for the catch.

Once we test conjugated inputs on all the possible inputs (similar to the extensions we did for `noncontiguous`), these samples will not be necessary.

anjali411 replied:
@mruberry they are generated automatically for all relevant tests I think. 

I added this logic manually to generate conjugated samples when required:

```
include_conjugated_inputs = op.test_conjugated_samples and dtype.is_complex
        samples = op.sample_inputs(device, dtype, requires_grad=_requires_grad, include_conjugated_inputs=include_conjugated_inputs)
```

I see that we are not generating conjugated samples in `test_noncontiguous_samples`!

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -325,6 +325,10 @@ def _getattr_qual(obj, name, default=_NOTHING):
 #   a "name".
+#   All the "sample_inputs" functions are invoked within a `torch.no_grad()`
+#   environment for efficiency and correctness. As such remember to set the the
+#   "requires_grad" flag on the inputs **after** performing any transformations
```
What's a little scary about this change is that we have no way of validating if the things that should require grad still do after this, right? Like that evaluation is just done manually?

Lezcano replied:
It is true that, if no input from an opinfo has `requires_grad=True`, then we throw, but it could be that there are some operations that return just one of the inputs with `requries_grad` but not all of them.

One way to check this would be to throw if for an input of floating point, none of the inputs has `requires_grad=True`. This might give a few false positives, but I think it might even uncover some previous errors we missed.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2117,9 +2105,9 @@ def sample_inputs_baddbmm(op_info, device, dtype, requires_grad, **kwargs):
-                args[0].transpose(-1, 1).detach().requires_grad_(requires_grad),
-                args=(args[1].transpose(-1, 1).conj().detach().requires_grad_(requires_grad),
-                      args[2].transpose(-1, 1).conj().detach().requires_grad_(requires_grad)),
+                args[0].transpose_(-1, 1),
```
No requires grad here?

Lezcano replied:
No, as this just modifies the strides of the tensor, but does not return a new tensor.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5263,29 +5233,19 @@ def c(t):
-    The input is generated as the itertools.product of 'batches' and 'ns'.
-    """
+    # QR is just well defined when the matrix is full rank
+    make_fullrank = make_fullrank_matrices_with_distinct_singular_values
```
haha, that's one way to shorten line length

Lezcano replied:
I should really put up a PR renaming this function...

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5263,29 +5233,19 @@ def c(t):
 
 
 def sample_inputs_linalg_qr(op_info, device, dtype, requires_grad=False, **kwargs):
```
Add a comment that this is also used for geqrf

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5412,10 +5372,9 @@ def make_arg(shape):
-            result.narrow(dim_select[0], 0, 1).narrow(dim_select[1], 1, 1).zero_()
-            result.narrow(dim_select[0], 2, 1).narrow(dim_select[1], 3, 1).zero_()
-            result.narrow(dim_select[0], 4, 1).narrow(dim_select[1], 3, 1).zero_()
+        result.narrow(dim_select[0], 0, 1).narrow(dim_select[1], 1, 1).zero_()
```
The result of `narrow()` on a tensor that requires grad is a tensor that requires grad even under the no grad context?

Lezcano replied:
The result of narrow is a new tensor which is a view on a part of a tensor, that then is zeroed out in place. As such, the original tensor object is not modified (although its values are)

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6017,103 +5974,50 @@ def sample_inputs_i0_i1(op_info, device, dtype, requires_grad, **kwargs):
 
-    int_alpha, float_alpha, complex_alpha = 2, 0.1, 1 + 0.6j
+    shapes = ((S, S), (S,), ())
+    if dtype.is_complex:
```
This needs a comment

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6017,103 +5974,50 @@ def sample_inputs_i0_i1(op_info, device, dtype, requires_grad, **kwargs):
-        return (SampleInput(input.detach().clone().requires_grad_(requires_grad),
-                            args=(arg,), kwargs=dict(alpha=alpha))
-                for (input, arg), alpha in filtered_product)
+def sample_inputs_rsub(op_info, device, dtype, requires_grad, other_scalar, **kwargs):
```
Don't we have generic elementwise binary sample inputs now for ops like add and sub? Can this just reuse that function?

Lezcano replied:
We do have `sample_inputs_rbinops`. Now, it seems rather simple. I wonder whether we should use a more involved scheme like hte one proposed here that tests for tensor-tensor / scalar-tensor operations with different dtypes and all. What do you think?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -13002,8 +12902,6 @@ def ref_pairwise_distance(input1, input2):
             DecorateInfo(unittest.skip("Skipped!"), 'TestReductions', 'test_dim_none'),
             DecorateInfo(unittest.skip("Skipped!"), 'TestReductions', 'test_dim_none_keepdim'),
             # FIXME: improve precision
-            DecorateInfo(unittest.skip("Skipped!"), 'TestReductions', 'test_noncontiguous_all',
```
Interesting... what fixed this?

Lezcano replied:
I believe it was failing due to some problems in the previous generation of the inputs (perhaps some flags of the final tensor were not set properly). After rewriting the sampling I run all these to see if it had fixed any of the problems and...

mruberry reviewed (commented):
Another super cool and interesting PR, @Lezcano. cc @saketh-are @pmeier @kshitij12345 and @krshrimali.

A few questions/comments inline for your review

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

Lezcano commented:
I just added the suggested changes / answered some of the questions @mruberry

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

Lezcano commented:
@pytorchbot ciflow rerun -l ciflow/all

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

Lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5297,27 +5283,18 @@ def sample_inputs_lu_solve(op_info, device, dtype, requires_grad=False, **kwargs
             b_.requires_grad_(b_grad)
             yield SampleInput(b_, args=(lu_, pivs))
 
+            lu_ = lu.detach().clone()
+            lu_.requires_grad_(lu_grad)
+            b_ = b.detach().clone()
+            b_.requires_grad_(b_grad)
+            yield SampleInput(b_, args=(lu_, pivs))
+
```
remove

mruberry reviewed (approved):
Cool! Just one duplicate code block

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774420).

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

Differential Revision: [D33774420](https://our.internmc.facebook.com/intern/diff/D33774420)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774420).

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

Differential Revision: [D33774420](https://our.internmc.facebook.com/intern/diff/D33774420)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774420).

lezcano committed:
Update on "Deactivate the tracking of gradients in sampling functions within OpInfos"


Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

Differential Revision: [D33774420](https://our.internmc.facebook.com/intern/diff/D33774420)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774420).
